[hw,dma,rtl] Lock configuration registers during an operation

Razer6 opened the pull request:
This PR implements a hardware managed REGWEN, that locks up the configuration registers of the DMA during an operation. When the DMA leaves the `DmaIdle` state, the hardware locks the following registers:

|  Register| Description |
| ------------- | ------------- |
| `SOURCE_ADDRESS_LO` | Locked  |
| `SOURCE_ADDRESS_HI` | Locked  |
| `DESTINATION_ADDRESS_LO` | Locked  |
| `DESTINATION_ADDRESS_HI` | Locked  |
| `ADDRESS_SPACE_ID` | Locked  |
| `ENABLED_MEMORY_RANGE_BASE` | Not locked but captured in the DMA since it already has a REGWEN  |
| `ENABLED_MEMORY_RANGE_LIMIT` | Not locked but captured in the DMA since it already has a REGWEN  |
| `RANGE_VALID` | Not locked but captured in the DMA since it already has a REGWEN  |
| `RANGE_REGWEN` | Not Locked  |
| `CFG_REGWEN` | Register that implements that lock.  |
| `TOTAL_DATA_SIZE` | Locked  |
| `CHUNK_DATA_SIZE` | Locked  |
| `TRANSFER_WIDTH` | Locked  |
| `DESTINATION_ADDRESS_LIMIT_LO` | Locked  |
| `DESTINATION_ADDRESS_LIMIT_HI` | Locked  |
| `DESTINATION_ADDRESS_ALMOST_LIMIT_LO` | Locked  |
| `DESTINATION_ADDRESS_ALMOST_LIMIT_HI` | Locked  |
| `CONTROL` | Opcode, HW handshake, *increment_enable, direction captured.  |
| `STATUS` | Not locked  |
| `SHA2_DIGEST` | Not locked, read only  |
| `HANDSHAKE_INTERRUPT_ENABLE` | Locked  |
| `CLEAR_INT_SRC` | Locked  |
| `CLEAR_INT_BUS` | Locked  |
| `INT_SOURCE_ADDR` | Locked  |
| `INT_SOURCE_WR_VAL` | Locked  |

alees24 reviewed (approved):
Thanks. I think this is okay, subject to resolving the following 3 points:

- 'hw2reg.control.go' has multiple assignments. The first assignment, with the abort term also, is required.
- pick up the 'status.busy.q' reset change already merged, because this is now critical to the controller's function.
- please be sure that 'busy.q' is also cleared by the abort operation, ie. use the logic in the first 'control.go' assignment and not just the simpler term 'clear_go'.

The commit history could perhaps be a littler simpler across the first 3 commits, rather than making changes in one and then modifying them in the next, eg. the capturing of asid values and the DMA-enabled memory range, but I realize that this is likely an effort to keep each commit small.

Razer6 commented:
> * 'hw2reg.control.go' has multiple assignments. The first assignment, with the abort term also, is required.

Removed.

> * pick up the 'status.busy.q' reset change already merged, because this is now critical to the controller's function.

Rebased.

> * please be sure that 'busy.q' is also cleared by the abort operation, ie. use the logic in the first 'control.go' assignment and not just the simpler term 'clear_go'.

Added `cfg_abort_en `. The `&& ctrl_state_d == DmaIdle` was removed as this is always the case.
